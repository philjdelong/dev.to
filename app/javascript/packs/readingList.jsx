import { h, render } from 'preact';
import { getUserDataAndCsrfToken } from '../chat/util';
import { ReadingList } from '../readingList/readingList';

function loadElement() {
  // fetch user information and auth token 
  getUserDataAndCsrfToken().then(({ currentUser }) => {
    // target div from reading list view
    const root = document.getElementById('reading-list');
    // render the ReadingList component with user data and data passed from div
    if (root) {
      render(
        <ReadingList
          availableTags={currentUser.followed_tag_names}
          statusView={root.dataset.view}
        />,
        root,
        root.firstElementChild,
      );
    }
  });
}

// utilize InstantClick to pre-render on hover
window.InstantClick.on('change', () => {
  loadElement();
});

loadElement();
