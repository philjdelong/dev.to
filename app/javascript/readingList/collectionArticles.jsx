import { h } from 'preact';
import PropTypes from 'prop-types';
import { Article } from './components/Article';

export const CollectionArticles = collection => {
  const articles = JSON.parse(collection.articles);
  const articleList = articles.map(article => {
    return <Article {...article} />;
  });

  return <section className="collection item-wrapper">{articleList}</section>;
};

CollectionArticles.propTypes = {
  collection: PropTypes.string.isRequired,
};
