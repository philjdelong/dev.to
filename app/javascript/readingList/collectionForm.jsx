import { h, Component } from 'preact';
import PropTypes from 'prop-types';
import linkState from 'linkstate';
import Tags from '../shared/components/tags';

export class CollectionForm extends Component {
  constructor({ tagList }) {
    super({ tagList });
    this.state = {
      // eslint-disable-next-line react/no-unused-state
      title: '',
      tagList,
      // auth: false
    };
  }

  handleChange = e => {
    this.setState({ [e.target.name]: e.target.value });
  };

  submitCollection = (e, title, tags) => {
    // check that tag length is > 1, title req handled on input
    // need to modify route also and ensure error handling is included.
    e.preventDefault();
    fetch('/reading_collections', {
      method: 'POST',
      headers: {
        Accept: 'application/json',
        'X-CSRF-Token': window.csrfToken,
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        title,
        tags,
      }),
      credentials: 'same-origin',
    });
  };

  render() {
    const { tagList } = this.state;

    return (
      <div className="articleformcontainer">
        <section className="articleform">
          <form className="articleform__form">
            <h1>Add your new collection</h1>
            <input
              type="text"
              name="title"
              placeholder="Title..."
              onChange={this.handleChange}
              className="articleform__title"
              required
            />
            <Tags
              defaultValue={tagList}
              onInput={linkState(this, 'tagList')}
              maxTags={4}
              autoComplete="off"
              classPrefix="articleform"
            />
            {/* <h4>{`${this.state.error}`}</h4> */}
            <a
              className="articleform__buttons--publish collection-btn"
              type="button"
              href="readinglist"
            >
              CREATE COLLECTION
            </a>
          </form>
        </section>
      </div>
    );
  }
}

CollectionForm.propTypes = {
  tagList: PropTypes.arrayOf(PropTypes.object).isRequired,
};
