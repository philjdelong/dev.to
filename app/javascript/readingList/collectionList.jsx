import { h, Component } from 'preact';
import PropTypes from 'prop-types';
import { Collection } from './components/Collection';

export class CollectionList extends Component {
  constructor({ collections }) {
    super({ collections });

    this.state = {
      collections,
    };
  }

  render() {
    const { collections } = this.state;
    const collectionsToRender = collections.map(collection => {
      return (
        <Collection
          key={collection.id}
          name={collection.name}
          tags={collection.tag_list}
          slug={collection.slug}
        />
      );
    });

    return (
      <section
        className="collection-cont results results--loaded"
        id="reading-collection"
      >
        <div className="results-header collection-header">
          {collections.length
            ? `Collections (${collections.length})`
            : 'Collections'}
          <a className="new-collection" href="/readingcollections/new">
            +
          </a>
        </div>
        {collectionsToRender}
      </section>
    );
  }
}

CollectionList.propTypes = {
  collections: PropTypes.arrayOf(PropTypes.object).isRequired,
};
