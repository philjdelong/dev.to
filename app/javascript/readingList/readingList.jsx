// This seems important but I'm not totally sure what it is doing! its our actual readinglist compoment!
// can we put a button to a get /readinglist/new here?
import { h, Component } from 'preact';
import { PropTypes } from 'preact-compat';
import debounce from 'lodash.debounce';
// SIL number of helper fuctions
// allowing for user interaction
// tag selection, searching, pagnation
import {
  defaultState,
  loadNextPage,
  onSearchBoxType,
  performInitialSearch,
  search,
  toggleTag,
  clearSelectedTags,
} from '../searchableItemList/searchableItemList';

// abstracts out other components that have been defined elsewhere
// makes the readingList less cluttered

import { ItemListItem } from '../src/components/ItemList/ItemListItem'; // this has details for each list item
import { ItemListItemArchiveButton } from '../src/components/ItemList/ItemListItemArchiveButton'; // archive button
import { ItemListLoadMoreButton } from '../src/components/ItemList/ItemListLoadMoreButton'; // load more button
import { ItemListTags } from '../src/components/ItemList/ItemListTags'; // tags

const STATUS_VIEW_VALID = 'valid';
const STATUS_VIEW_ARCHIVED = 'archived';
const READING_LIST_ARCHIVE_PATH = '/readinglist/archive';
const READING_LIST_PATH = '/readinglist';

const FilterText = ({ selectedTags, query, value }) => {
  return (
    <h1>
      {selectedTags.length === 0 && query.length === 0
        ? value
        : 'Nothing with this filter ðŸ¤”'}
    </h1>
  );
};
// reading list class
export class ReadingList extends Component {
  // defines the state / where data is being stored in the component
  constructor(props) {
    super(props);
    // keeps track of user tags and status
    // this is how data gets passed from the prop
    const { availableTags, statusView } = this.props;

    // imports state from global state management
    // defaultState is coming from searchable item list
    this.state = defaultState({ availableTags, archiving: false, statusView });

    // bind and initialize all shared functions
    this.onSearchBoxType = debounce(onSearchBoxType.bind(this), 300, {
      leading: true,
    });
    this.loadNextPage = loadNextPage.bind(this);
    this.performInitialSearch = performInitialSearch.bind(this);
    this.search = search.bind(this);
    this.toggleTag = toggleTag.bind(this);
    this.clearSelectedTags = clearSelectedTags.bind(this);
  }

  // when a component mounts, load their reading list times
  componentDidMount() {
    const { hitsPerPage, statusView } = this.state;
    // this is going to angolia? and where the article data is coming from
    // kind of like an endpoint hit / fetch
    this.performInitialSearch({
      containerId: 'reading-list', // used in the setup angolia index, essentially setting up context of the search
      indexName: 'SecuredReactions',
      searchOptions: {
        hitsPerPage,
        filters: `status:${statusView}`,
      },
    });
  }

  toggleStatusView = event => {
    event.preventDefault();

    const { query, selectedTags } = this.state;

    const isStatusViewValid = this.statusViewValid();
    const newStatusView = isStatusViewValid
      ? STATUS_VIEW_ARCHIVED
      : STATUS_VIEW_VALID;
    const newPath = isStatusViewValid
      ? READING_LIST_ARCHIVE_PATH
      : READING_LIST_PATH;

    // empty items so that changing the view will start from scratch
    this.setState({ statusView: newStatusView, items: [] });

    this.search(query, {
      page: 0,
      tags: selectedTags,
      statusView: newStatusView,
    });

    // change path in the address bar
    window.history.replaceState(null, null, newPath);
  };

  toggleArchiveStatus = (event, item) => {
    event.preventDefault();

    const { statusView, items, totalCount } = this.state;
    // fetch request happening here!
    // need to find the backend test for this
    window.fetch(`/reading_list_items/${item.id}`, {
      method: 'PUT',
      headers: {
        'X-CSRF-Token': window.csrfToken,
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({ current_status: statusView }),
      credentials: 'same-origin',
    });

    const t = this;
    const newItems = items;
    newItems.splice(newItems.indexOf(item), 1);
    t.setState({
      archiving: true,
      items: newItems,
      totalCount: totalCount - 1,
    });
    // whats a snack bar? some sort of a nav bar?
    // hide the snackbar in a few moments
    setTimeout(() => {
      t.setState({ archiving: false });
    }, 1000);
  };

  statusViewValid() {
    const { statusView } = this.state;
    return statusView === STATUS_VIEW_VALID;
  }

  renderEmptyItems() {
    const { itemsLoaded, selectedTags, query } = this.state;

    if (itemsLoaded && this.statusViewValid()) {
      return (
        <div className="items-empty">
          <FilterText
            selectedTags={selectedTags}
            query={query}
            value="Your Reading List is Lonely" // displays if reading list is empty
          />
          <h3>
            Hit the
            <span className="highlight">SAVE</span>
            or
            <span className="highlight">
              Bookmark
              <span role="img" aria-label="Bookmark">
                ðŸ”–
              </span>
            </span>
            to start your Collection
          </h3>
        </div>
      );
    }

    return (
      <div className="items-empty">
        <FilterText
          selectedTags={selectedTags}
          query={query}
          value="Your Archive List is Lonely"
        />
      </div>
    );
  }
  // imortant to see where the data ends up
  render() {
    const {
      // destructuring from an objects
      // these are like attributes of the state object
      items, // items are set in state
      itemsLoaded,
      totalCount,
      availableTags,
      selectedTags,
      showLoadMoreButton,
      archiving,
    } = this.state;

    const isStatusViewValid = this.statusViewValid();
    // button to archive vs un archive?
    const archiveButtonLabel = isStatusViewValid ? 'archive' : 'unarchive';
    // this must be the reading list items
    const itemsToRender = items.map(item => {
      return (
        <ItemListItem item={item}>
          <ItemListItemArchiveButton
            text={archiveButtonLabel}
            onClick={e => this.toggleArchiveStatus(e, item)}
          />
        </ItemListItem>
      );
    });

    const snackBar = archiving ? (
      <div className="snackbar">
        {isStatusViewValid ? 'Archiving...' : 'Unarchiving...'}
      </div>
    ) : (
      ''
    );
    return (
      // this is the sidebar I was not able to locate before
      <div className="home item-list">
        <div className="side-bar">
          <div className="widget filters">
            <input
              onKeyUp={this.onSearchBoxType}
              placeHolder="search your list"
            />
            <div className="filters-header">
              <h4 className="filters-header-text">my tags</h4> // the users tags
              have been found!
              {Boolean(selectedTags.length) && (
                <a
                  className="filters-header-action"
                  href={
                    isStatusViewValid
                      ? READING_LIST_PATH
                      : READING_LIST_ARCHIVE_PATH
                  }
                  onClick={this.clearSelectedTags}
                  data-no-instant
                >
                  clear all
                </a>
              )}
            </div>
            <ItemListTags
              availableTags={availableTags}
              selectedTags={selectedTags}
              onClick={this.toggleTag}
            />

            <div className="status-view-toggle">
              <a
                href={READING_LIST_ARCHIVE_PATH}
                onClick={e => this.toggleStatusView(e)}
                data-no-instant
              >
                {isStatusViewValid ? 'View Archive' : 'View Reading List'}
              </a>
            </div>
          </div>
        </div>

        <div className="items-container">
          {' '}
          // next section is the box where all of the reading list items are
          <div className={`results ${itemsLoaded ? 'results--loaded' : ''}`}>
            <div className="results-header">
              {isStatusViewValid ? 'Reading List' : 'Archive'}
              {` (${totalCount > 0 ? totalCount : 'empty'})`} // total count of
              the items in the reading list
            </div>
            <div>
              {items.length > 0 ? itemsToRender : this.renderEmptyItems()}
            </div>
          </div>
          <ItemListLoadMoreButton
            show={showLoadMoreButton}
            onClick={this.loadNextPage}
          />
        </div>

        {snackBar}
      </div>
    );
  }
}

ReadingList.defaultProps = {
  statusView: STATUS_VIEW_VALID,
};

ReadingList.propTypes = {
  availableTags: PropTypes.arrayOf(PropTypes.string).isRequired,
  statusView: PropTypes.oneOf([STATUS_VIEW_VALID, STATUS_VIEW_ARCHIVED]),
};

FilterText.propTypes = {
  selectedTags: PropTypes.arrayOf(PropTypes.string).isRequired,
  value: PropTypes.string.isRequired,
  query: PropTypes.arrayOf(PropTypes.string).isRequired,
};
