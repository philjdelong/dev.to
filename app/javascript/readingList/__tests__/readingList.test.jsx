import { h } from 'preact';
import render from 'preact-render-to-json';
import { ReadingList } from '../readingList';

describe('<ReadingList />', () => {
  it('renders properly', () => {
    const mockCollections = [
      { id: 1, name: 'first_test_post' },
      { id: 2, name: 'second_test_post' },
    ];
    const tree = render(
      <ReadingList availableTags={['discuss']} collections={mockCollections} />,
    );
    expect(tree).toMatchSnapshot();
  });
});
