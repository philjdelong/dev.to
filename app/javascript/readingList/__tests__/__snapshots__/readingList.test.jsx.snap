// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ReadingList /> renders properly 1`] = `
<div
  class="home item-list"
>
  <div
    class="side-bar"
  >
    <div
      class="widget filters"
    >
      <input
        onKeyUp={[Function]}
        placeHolder="search your list"
      />
      <div
        class="filters-header"
      >
        <h4
          class="filters-header-text"
        >
          my tags
        </h4>
      </div>
      <div
        class="tags"
      >
        <a
          class="tag "
          data-no-instant={true}
          href="/t/discuss"
          onClick={[Function]}
        >
          #discuss
        </a>
      </div>
      <div
        class="status-view-toggle"
      >
        <a
          data-no-instant={true}
          href="/readinglist/archive"
          onClick={[Function]}
        >
          View Archive
        </a>
      </div>
    </div>
  </div>
  <div
    class="items-container"
  >
    <div
      class="results "
    >
      <div
        class="results-header"
      >
        Reading List (empty)
      </div>
      <div>
        <div
          class="items-empty"
        >
          <h1>
            Your Archive List is Lonely
          </h1>
        </div>
      </div>
    </div>
  </div>
  <div
    class="items-container"
  >
    <div
      class="collection-cont results results--loaded"
      id="reading-collection"
    >
      <div
        class="results-header collection-header"
      >
        Collections (2)
        <a
          class="new-collection"
          href="/readingcollections/new"
        >
          +
        </a>
      </div>
      <section
        class="collection item-wrapper"
      >
        <div
          class="item"
        >
          <a
            class="item-title"
            href="/readingcollections/testing-collection-2ioh"
          >
            Testing Collection
          </a>
          <section
            class="tag-display"
          >
            <a
              href="/t/javascript"
            >
              <span
                class="tag"
              >
                 #javascript
              </span>
            </a>
            <a
              href="/t/webdev"
            >
              <span
                class="tag"
              >
                 #webdev
              </span>
            </a>
          </section>
        </div>
      </section>
      <section
        class="collection item-wrapper"
      >
        <div
          class="item"
        >
          <a
            class="item-title"
            href="/readingcollections/testing-collection2-2dwa"
          >
            Testing Collection 2
          </a>
          <section
            class="tag-display"
          >
            <a
              href="/t/security"
            >
              <span
                class="tag"
              >
                 #security
              </span>
            </a>
            <a
              href="/t/java"
            >
              <span
                class="tag"
              >
                 #java
              </span>
            </a>
          </section>
        </div>
      </section>
    </div>
  </div>
</div>
`;
